pybamm/expression_tree/jacobian.py

```diff
@@ -0,0 +1,92 @@
+#
+# Calculate the Jacobian of a symbol
+#
+import pybamm
+
+
+class Jacobian(object):
+    def __init__(self, known_jacs=None):
+        self._known_jacs = known_jacs or {}
+
+    def jac(self, symbol, variable):
+        """
+        This function recurses down the tree, computing the Jacobian using
+        the Jacobians defined in classes derived from pybamm.Symbol. E.g. the
+        Jacobian of a 'pybamm.Multiplication' is computed via the product rule.
+        If the Jacobian of a symbol has already been calculated, the stored value
+        is returned.
+        Note: The Jacobian is the derivative of a symbol with respect to a (slice of)
+        a State Vector.
+
+        Parameters
+        ----------
+        symbol : :class:`pybamm.Symbol`
+            The symbol to calculate the Jacobian of
+        variable : :class:`pybamm.Symbol`
+            The variable with respect to which to differentiate
+
+        Returns
+        -------
+        :class:`pybamm.Symbol`
+            Symbol representing the Jacobian
+        """
+
+        try:
+            return self._known_jacs[symbol.id]
+        except KeyError:
+            jac = self._jac(symbol, variable)
+            self._known_jacs[symbol.id] = jac
+            return jac
+
+    def _jac(self, symbol, variable):
+        """ See :meth:`Jacobian.jac()`. """
+
+        if isinstance(symbol, pybamm.Outer):
+            # Need to treat outer differently. If the left child of an Outer
+            # evaluates to number then we need to return a matrix of zeros
+            # of the correct size, which requires variable.evaluation_array
+            left, right = symbol.children
+            # process children
+            left_jac = self.jac(left, variable)
+            right_jac = self.jac(right, variable)
+            # _outer_jac defined in pybamm.Outer
+            jac = symbol._outer_jac(left_jac, right_jac, variable)
```

Comment:
not a big deal but you could have the isinstance(Outer) check on line 60 to avoid writing the process_children lines twice